File: modal-js/src/secret.ts
import { getDefaultClient, type ModalClient } from "./client";
import { ClientError, Status } from "nice-grpc";
import { InvalidError, NotFoundError } from "./errors";
import { ObjectCreationType } from "../proto/modal_proto/api";

/** Optional parameters for {@link SecretService#fromName client.secrets.fromName()}. */
export type SecretFromNameParams = {
  environment?: string;
  requiredKeys?: string[];
};

/** Optional parameters for {@link SecretService#fromObject client.secrets.fromObject()}. */
export type SecretFromObjectParams = {
  environment?: string;
};

/**
 * Service for managing {@link Secret Secrets}.
 *
 * Normally only ever accessed via the client as:
 * ```typescript
 * const modal = new ModalClient();
 * const secret = await modal.secrets.fromName("my-secret");
 * ```
 */
export class SecretService {
  readonly #client: ModalClient;
  constructor(client: ModalClient) {
    this.#client = client;
  }

  /** Reference a {@link Secret} by its name. */
  async fromName(name: string, params?: SecretFromNameParams): Promise<Secret> {
    try {
      const resp = await this.#client.cpClient.secretGetOrCreate({
        deploymentName: name,
        environmentName: this.#client.environmentName(params?.environment),
        requiredKeys: params?.requiredKeys ?? [],
      });
      return new Secret(resp.secretId, name);
    } catch (err) {
      if (err instanceof ClientError && err.code === Status.NOT_FOUND)
        throw new NotFoundError(err.details);
      if (
        err instanceof ClientError &&
        err.code === Status.FAILED_PRECONDITION &&
        err.details.includes("Secret is missing key")
      )
        throw new NotFoundError(err.details);
      throw err;
    }
  }

  /** Create a {@link Secret} from a plain object of key-value pairs. */
  async fromObject(
    entries: Record<string, string>,
    params?: SecretFromObjectParams,
  ): Promise<Secret> {
    for (const [, value] of Object.entries(entries)) {
      if (value == null || typeof value !== "string") {
        throw new InvalidError(
          "entries must be an object mapping string keys to string values, but got:\n" +
            JSON.stringify(entries),
        );
      }
    }

    try {
      const resp = await this.#client.cpClient.secretGetOrCreate({
        objectCreationType: ObjectCreationType.OBJECT_CREATION_TYPE_EPHEMERAL,
        envDict: entries as Record<string, string>,
        environmentName: this.#client.environmentName(params?.environment),
      });
      return new Secret(resp.secretId);
    } catch (err) {
      if (
        err instanceof ClientError &&
        (err.code === Status.INVALID_ARGUMENT ||
          err.code === Status.FAILED_PRECONDITION)
      )
        throw new InvalidError(err.details);
      throw err;
    }
  }
}

/** Secrets provide a dictionary of environment variables for {@link Image}s. */
export class Secret {
  readonly secretId: string;
  readonly name?: string;

  /** @ignore */
  constructor(secretId: string, name?: string) {
    this.secretId = secretId;
    this.name = name;
  }

  /**
   * @deprecated Use {@link SecretService#fromName client.secrets.fromName()} instead.
   */
  static async fromName(
    name: string,
    params?: SecretFromNameParams,
  ): Promise<Secret> {
    return getDefaultClient().secrets.fromName(name, params);
  }

  /**
   * @deprecated Use {@link SecretService#fromObject client.secrets.fromObject()} instead.
   */
  static async fromObject(
    entries: Record<string, string>,
    params?: SecretFromObjectParams,
  ): Promise<Secret> {
    return getDefaultClient().secrets.fromObject(entries, params);
  }
}

export async function mergeEnvIntoSecrets(
  client: ModalClient,
  env?: Record<string, string>,
  secrets?: Secret[],
): Promise<Secret[]> {
  const result = [...(secrets || [])];
  if (env && Object.keys(env).length > 0) {
    result.push(await client.secrets.fromObject(env));
  }
  return result;
}

File: modal-go/secret.go
package modal

import (
	"context"

	pb "github.com/modal-labs/libmodal/modal-go/proto/modal_proto"
)

// SecretService provides Secret related operations.
type SecretService interface {
	FromName(ctx context.Context, name string, params *SecretFromNameParams) (*Secret, error)
	FromMap(ctx context.Context, keyValuePairs map[string]string, params *SecretFromMapParams) (*Secret, error)
}

type secretServiceImpl struct{ client *Client }

// Secret represents a Modal Secret.
type Secret struct {
	SecretID string
	Name     string
}

// SecretFromNameParams are options for finding Modal Secrets.
type SecretFromNameParams struct {
	Environment  string
	RequiredKeys []string
}

// FromName references a Secret by its name.
func (s *secretServiceImpl) FromName(ctx context.Context, name string, params *SecretFromNameParams) (*Secret, error) {
	if params == nil {
		params = &SecretFromNameParams{}
	}

	resp, err := s.client.cpClient.SecretGetOrCreate(ctx, pb.SecretGetOrCreateRequest_builder{
		DeploymentName:  name,
		EnvironmentName: environmentName(params.Environment, s.client.profile),
		RequiredKeys:    params.RequiredKeys,
	}.Build())

	if err != nil {
		return nil, err
	}

	return &Secret{SecretID: resp.GetSecretId(), Name: name}, nil
}

// SecretFromMapParams are options for creating a Secret from a key/value map.
type SecretFromMapParams struct {
	Environment string
}

// FromMap creates a Secret from a map of key-value pairs.
func (s *secretServiceImpl) FromMap(ctx context.Context, keyValuePairs map[string]string, params *SecretFromMapParams) (*Secret, error) {
	if params == nil {
		params = &SecretFromMapParams{}
	}

	resp, err := s.client.cpClient.SecretGetOrCreate(ctx, pb.SecretGetOrCreateRequest_builder{
		ObjectCreationType: pb.ObjectCreationType_OBJECT_CREATION_TYPE_EPHEMERAL,
		EnvDict:            keyValuePairs,
		EnvironmentName:    environmentName(params.Environment, s.client.profile),
	}.Build())
	if err != nil {
		return nil, err
	}
	return &Secret{SecretID: resp.GetSecretId()}, nil
}

// mergeEnvIntoSecrets merges environment variables into the secrets list.
// If env contains values, it creates a new Secret from the env map and appends it to the existing secrets.
func mergeEnvIntoSecrets(ctx context.Context, client *Client, env *map[string]string, secrets *[]*Secret) ([]*Secret, error) {
	var result []*Secret
	if secrets != nil {
		result = append(result, *secrets...)
	}

	if env != nil && len(*env) > 0 {
		envSecret, err := client.Secrets.FromMap(ctx, *env, nil)
		if err != nil {
			return nil, err
		}
		result = append(result, envSecret)
	}

	return result, nil
}
